llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp

```diff
@@ -254,33 +256,55 @@ PreservedAnalyses CompileTimePropertiesPass::run(Module &M,
     }
   }
 
-  // Process all properties on kernels.
   for (Function &F : M) {
     // Only consider kernels.
     if (F.getCallingConv() != CallingConv::SPIR_KERNEL)
       continue;
 
-    SmallVector<Metadata *, 8> MDOps;
-    SmallVector<std::pair<std::string, MDNode *>, 8> NamedMDOps;
-    for (const Attribute &Attribute : F.getAttributes().getFnAttrs()) {
-      if (MDNode *SPIRVMetadata = attributeToDecorateMetadata(Ctx, Attribute))
-        MDOps.push_back(SPIRVMetadata);
-      else if (auto NamedMetadata = attributeToExecModeMetadata(M, Attribute))
-        NamedMDOps.push_back(*NamedMetadata);
+    {
+      // Process all properties on kernels arugments
+      SmallVector<Metadata *, 8> MDOps;
+      for (unsigned i = 0; i < F.arg_size(); i++) {
+        SmallVector<Metadata *, 8> MDArgOps;
+        for (auto &Attribute : F.getAttributes().getParamAttrs(i)) {
+          if (MDNode *SPIRVMetadata =
+                  attributeToDecorateMetadata(Ctx, Attribute))
+            MDArgOps.push_back(SPIRVMetadata);
+        }
+        MDOps.push_back(MDNode::get(Ctx, MDArgOps));
+      }
+      // Add the generated metadata to the kernel function.
+      if (!MDOps.empty()) {
```

Comment:
Seems like this will only be empty if `F.arg_size() == 0`. Although that should be fine, we could reduce the clutter a little by not adding it if none of the arguments had SPIRV decoration metadata. The simplest solution would be to just use a boolean to track whether or not we found any with `attributeToDecorateMetadata`.

Reply:
Fixed

Reply:
Is the fix the 
```c++
if (!MDArgOps.empty())
  MDOps.push_back(MDNode::get(Ctx, MDArgOps));
```
above? If so, would it not cause problems when only select arguments are missing the attribute? That is, when one-or-more function arguments have recognized attributes we need to have an entry in `MDOps` for each of them so there is a 1:1 mapping.

What I was thinking was something along the lines of having a bool you set to true if your new if-statement, then use that instead of `!MDOps.empty()` here. To be specific you would have something like:

```c++
      bool ArgWithMDMet = false;
      for (unsigned i = 0; i < F.arg_size(); i++) {
        SmallVector<Metadata *, 8> MDArgOps;
        for (auto &Attribute : F.getAttributes().getParamAttrs(i)) {
          if (MDNode *SPIRVMetadata =
                  attributeToDecorateMetadata(Ctx, Attribute))
            MDArgOps.push_back(SPIRVMetadata);
        }
        if (!MDArgOps.empty())
          ArgWithMDMet = true;
        MDOps.push_back(MDNode::get(Ctx, MDArgOps));
      }
      // Add the generated metadata to the kernel function.
      if (ArgWithMDMet) {
```

Reply:
Changed. Sorry I wasn't thinking previously